Truncate log file during operation

pinheadmz opened the pull request:
I'm not sure if this is exactly the best approach but something needs to be done to this package to prevent DoS attacks by disk-filling the log. Recently on the Handshake network a few of us with `log-level: spam` were crashed by a pathological node on the network that repeatedly broadcast 10,000 orphan transactions. Currently `blgr` only truncates the log file during `open()`. Meaning if you are lucky enough never to have to restart your node, the log file will never stop growing.

pinheadmz committed:
truncate log during operation

Co-authored-by: Braydon Fuller <courier@braydon.com>

braydonf commented:
It should be possible to keep the size in memory and increment when writing. This will avoid the need to stat the file for every log statement. Truncate can then do a simple comparison and return early.

braydonf commented:
Also note: Another mitigation is to disable logging to file (`debug.log`) and redirect standard output to a logger such as systemd or similar via a service. It will handle log rotation and aggregation and will also be preferrable for [other reasons](https://github.com/bcoin-org/bcoin/issues/747).

pinheadmz commented:
@braydonf added a new commit to this patch that keeps a counter in memory and only triggers the truncate function when that counter exceeds a threshold. Due to the synchronous nature of logger, I rewrote the truncate function to run synchronously `truncateSync()` which is unfortunately required for this to work (I think). Note that the original `truncate()` function is the only method that used all the `promisify` stuff in the file `fs.js`. If we think this approach is OK we can remove the promisifying `fs.js` file and just call the nodejs filesystem functions synchronously, as mother nature intended. The original `truncate()` function with all its async calls promisified is only run once, at `open()` anyway -- that could be replaced the `truncateSync()` function I just added.

Going to pull this branch into a full node and let 'er rip.

braydonf commented:
> Due to the synchronous nature of logger, I rewrote the truncate function to run synchronously truncateSync() which is unfortunately required for this to work (I think).

The other option could be for `trucate()` to be `async`, however `truncate()` would not run again if there was a pending truncate call.

pinheadmz commented:
I had issues testing this solution with truncate being async -- all the functions that might call it are synchronous `logger.debug('...')` and so forth. I looked into `setInterval` even, but that's messy with promises. Even when iI shoved it into the synchronous functions, the file was being written and truncated in the wrong order and I ended up with huge test files anyway.

braydonf commented:
> all the functions that might call it are synchronous logger.debug('...')

Yes, however logging does not need to wait for the truncate to complete, so it can still call an async function and not wait for it complete. Doing that could lead to a case where there are multiple parallel calls to `truncate()`. A check can be added to avoid that, see https://github.com/braydonf/blgr/commit/fda1fceddc91a5d90def8a3ef41bfd7f15e4dcae.

pinheadmz commented:
@braydonf Rebased and squashed with your commit. Performance seems good. After syncing hsd with a `MAX_FILE_SIZE` set very low (`1 << 20`), I noticed some messy lines in the log file (below). 

Some lines are cut in half, which I'm not sure it's worth trying to fix (the `truncate()` function could scan for a `\n` to splice on?) 

The repeated "Truncating..." message might be due to the fact that `truncate()` prints that message to the log, then copies the entire log into memory before truncating the actual file. The reason for this is, I think, because `fs.ftruncate` cuts off the _end_ of the file, when our desire is to cut off the beginning.

```
[I:2019-10-10T12:01:14Z] (chain) Block b78501c09cebaf88e906f7f519088f539904468d4bc8676eaece006622b32492 (4066) added to chain (size=68942 txs=320 time=51.257049).
[D:2019-10-10T12:01:14Z] Truncating log file to 1mb.
[D:2019-10-10T12:01:14Z] Truncating log file to 1mb.
[D:2019-10-10T12:01:14Z] Truncating log file to 1mb.
b964f278d4871de642ef011ddf938e8a068b5 (4572) added to chain (size=570 txs=2 time=11.098214).
[D:2019-10-10T12:01:19Z] (chain) Retargetting to: 0000000206560000000000000000000000000000000000000000000000000000 (0x1d020656).
[D:2019-10-10T12:01:19Z] Truncating log file to 1mb.
[D:2019-10-10T12:01:19Z] (wallet) Adding block: 4573.
[D:2019-10-10T12:01:19Z] (chain) Memory: rss=156mb, js-heap=37/75mb native-heap=80mb
[I:2019-10-10T12:01:19Z] (chain) Block 47a96ee5a0a3d3cbe77c7402aaa394ef552c4bc55ac54149008d2bccdb1b30ac (4573) added to chain (size=355 txs=1 time=1.029639).
[D:2019-10-10T12:01:19Z] (chain) Retargetting to: 00000002044c0000000000000000000000000000000000000000000000000000 (0x1d02044c).
[D:2019-10-10T12:01:19Z] Truncating log file to 1mb.
ting log file to 1mb.
[D:2019-10-10T12:01:40Z] (wallet) Adding block: 6934.
[D:2019-10-10T12:01:40Z] (chain) Memory: rss=197mb, js-heap=48/87mb native-heap=110mb
[I:2019-10-10T12:01:40Z] (chain) Block c3a719b0cc106f64acac33b93c49f918fee68bd4f89745e0169ab93fef3c689f (6934) added to chain (size=355 txs=1 time=1.124365).
[D:2019-10-10T12:01:40Z] (chain) Retargetting to: 0000000b4e8f0000000000000000000000000000000000000000000000000000 (0x1d0b4e8f).
238) added to chain (size=355 txs=1 time=0.623967).
om peer with getblocks (138.68.61.31:13038).
[D:2019-10-10T12:02:14Z] (peer) Sending getblocks (hash=ecd23960f8dd0f8d5c98104e03135f2ded8fd1caa28bd71f5d27d8c37480fa92, stop=9da9015dbfa76d16bfe6f031b82d4aad3319235a73ebb06c5ce05c5444720f88).
```

pinheadmz committed:
prettier truncate

pinheadmz commented:
Added another commit that ensures truncated log is cut on newlines, and removed the "Truncating..." message, since we are in the truncating process, and writing a line to the log during that time results in unpredictable behavior. (do we need it?)

braydonf commented on lib/logger.js:
```diff
@@ -468,6 +487,9 @@ class Logger {
    */
 
   writeStream(level, module, args) {
+    if (this.truncating)
```
This was necessary? New logs should be added, and then previous logs removed correct?

pinheadmz replied:
It might not be necessary. I was trying to clear out issues where half of a line was stuck in the middle of the log, sometimes with the newline cutoff making the previous message long and weird. Especially with async `truncate`, I think it races against `writeStream`

braydonf commented:
An alternative here is to rotate logs and then unlink when necessary. For example `debug.log` would become `debug-2019-10-10.log.tar.gz` (or similar) and a new blank `debug.log` would be written. The unlinking could be done by another separate archiving process as the files would not be touched anymore.

tynes commented on lib/logger.js:
```diff
@@ -148,13 +151,16 @@ class Logger {
 
-    assert(!this.stream);
+    this.truncating = true;
 
```
Reading through the control flow:

`truncating` is set to true on line 163
It is then set to false in the `finally` clause on line 173
Will `truncating` ever be `true` on line 180? Reading the control flow, it doesn't look like it. Is setting it to `false` necessary?

pinheadmz replied:
huh good catch, I'll try taking it out

braydonf replied:
That should probably be inside the catch instead.

pinheadmz replied:
updated.

pinheadmz commented:
I've been testing this in hsd as well, syncing the chain is pretty quick and I manually set the MAX file size to 1 MB to watch it truncate.

braydonf commented:
With truncating at the beginning of the file, I think will need to read and move the data to the beginning. A benefit of rotating is that it's just moving the file and creating a new one. Writing would need to wait for the file to be open.

pinheadmz commented:
@braydonf are you thinking we should just abandon this approach and start a new PR that rotated log files? We can mitigate disk-fill by deleting old files after a maximum # of files have been created.

braydonf commented:
I don't know if it needs to be a new PR, just that rotating would solve the problem. I don't think it's more complicated, and may be more simple.

pinheadmz commented:
closing in favor of #8
